0009-api-v4-redesign.md

```diff
@@ -0,0 +1,212 @@
+
+One of the resources under `/me` of note is `/me/sessions`. This is used to handle logging in, logging out, and viewing all of the sessions of the currently logged in user.
+
+Some resources already more or less had names that are nouns also got names changes. One such change it calling private messages private messages `/direct-messages`. I replaced "private" with "direct" to better communicate that while these messages are sent directly to other users, they aren't private al a Matrix DMs. Another change is that what used to be referred to as "images" in the API is now "media". This change is partly to make it consistent with the list media endpoints from v3, and partly because it is technically also possible to upload non-images (like videos) with the old images endpoint. One more change is using the term "suspended" instead "remove" when referring to posts, comments, and communities that get taken down by moderators. This is because I wanted to make it clear that content that gets removed isn't necessarily removed permanently since mods can reinstate it. I'm not super happy with "suspended" and am open to ideas on alternatives.
+
+There were some other name changes to this effect, but the ones listed above are likely the most jarring ones.
+
+### PUT is Sometimes Used for Creation as well as Updates
+
+Some endpoints may seem odd at first glance because of this. Isn't PUT supposed to be used for updates? It is, but it can also be used for creating resources under certain circumstances. Recall the store resource archetype mentioned earlier: it says that URIs are chosen by the client instead of the server. What does this mean? An example will help make this clear.
+
+Take the path `/posts/locked/{postId}`. This uses PUT to lock a post, and DELETE to unlock a post. Note that the last part of the path is `postId`. `postId` is the ID of a post that already exists. In order to make the call, the client must already have the ID of a post in mind to lock. With this in mind, the PUT and DELETE operations on this endpoint can be though of more as a box that clients can put posts into and remove posts from.
+
+A major benefit of this approach is that it communicates that the operation is idempotent. A client could call `/posts/locked/1234` multiple times, and the effect would be the same as if they called it only a single time: lock a post if it's not already locked. Compare this to the v3 API, where locking a post requires POSTing to `/post/lock`. Even though the actual operation is idempotent, the POST method in HTTP is classified as not being idempotent; this results in some miscommunication about the API. PUT, on the other hand, is classified as being idempotent.
```

Comment:
How is this thing about idempotency useful? Does it actually cause different behavior?

Reply:
It doesn't cause different behavior, but it does communicate to the user of the API what to expect.